src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java

```diff
@@ -91,8 +96,77 @@ public void failed(Throwable exc, Void v) {
+        // Note that we don't include the cache header overhead in this calculation for simplicity,
+        // so putting entries which are just below the threshold may still cause this churn.
+        if (mTotalSize + entry.data.length > mMaxCacheSizeInBytes
+                && entry.data.length
+                        > mMaxCacheSizeInBytes * DiskBasedCacheUtility.HYSTERESIS_FACTOR) {
+            return;
+        }
+        final File file = DiskBasedCacheUtility.getFileForKey(key, mRootDirectorySupplier);
+        Path path = Paths.get(file.getPath());
+
+        try (AsynchronousFileChannel afc =
+                AsynchronousFileChannel.open(path, StandardOpenOption.WRITE)) {
+            final CacheHeader header = new CacheHeader(key, entry);
+            int headerSize = header.getHeaderSize();
+            int size = entry.data.length + headerSize;
+            ByteBuffer buffer = ByteBuffer.allocate(size);
+            header.writeHeader(buffer);
+            buffer.put(entry.data);
+            afc.write(
+                    /* source= */ buffer,
+                    /* position= */ 0,
+                    /* attachment= */ null,
+                    new CompletionHandler<Integer, Void>() {
+                        @Override
+                        public void completed(Integer integer, Void aVoid) {
+                            header.size = file.length();
+                            mTotalSize =
+                                    DiskBasedCacheUtility.putEntry(
+                                            key, header, mTotalSize, mEntries);
+                            mTotalSize =
+                                    DiskBasedCacheUtility.pruneIfNeeded(
+                                            mTotalSize,
+                                            mMaxCacheSizeInBytes,
+                                            mEntries,
+                                            mRootDirectorySupplier);
+                            callback.onPutComplete();
+                        }
+
+                        @Override
+                        public void failed(Throwable throwable, Void aVoid) {
+                            VolleyLog.e(
+                                    throwable, "Failed to read file %s", file.getAbsolutePath());
+                            callback.onPutComplete();
+                        }
+                    });
+        } catch (IOException e) {
+            boolean deleted = file.delete();
+            if (!deleted) {
+                VolleyLog.d("Could not clean up file %s", file.getAbsolutePath());
+            }
+            initializeIfRootDirectoryDeleted();
+            callback.onPutComplete();
+        }
+    }
+
+    public void initialize() {
+        // TODO (sphill99): #181
+    }
+
+    /** Re-initialize the cache if the directory was deleted. */
+    private void initializeIfRootDirectoryDeleted() {
+        if (!mRootDirectorySupplier.get().exists()) {
```

Comment:
nit: what do you think of returning early instead? I find that it improves readability as there is less nested code

```java
if (mRootDirectorySupplier.get().exists()) {
  return;
}

// Else do your work here
```

Reply:
fixed